## Assignment: $k$ Means Clustering

## **Do two questions.**

`! git clone https://www.github.com/DS3001/kmc`

**Q1.** This question is a case study for $k$ means clustering.

1. Load the `airbnb_hw.csv` data. Clean `Price` along with `Beds`, `Number Of Reviews`, and `Review Scores Rating`.

In [82]:
import numpy as np
import pandas as pd

airbnb = pd.read_csv('airbnb_hw.csv')
airbnb['N_Reviews'] = airbnb['Number Of Reviews']
airbnb['Score'] = airbnb['Review Scores Rating']
airbnb = airbnb.loc[:, ['Price', 'Beds', 'N_Reviews', 'Score']]
print(airbnb.shape)

(30478, 4)


In [83]:
print(airbnb.loc[airbnb['Price'] == ' '], '\n')
print(airbnb.loc[airbnb['Price'] == np.nan], '\n')
print(airbnb['Price'].value_counts())
airbnb['Price'] = airbnb['Price'].str.replace(',','')
airbnb['Price'] = pd.to_numeric(airbnb['Price'],errors='coerce')
print(airbnb['Price'].describe())

Empty DataFrame
Columns: [Price, Beds, N_Reviews, Score]
Index: [] 

Empty DataFrame
Columns: [Price, Beds, N_Reviews, Score]
Index: [] 

Price
150      1481
100      1207
200      1059
125       889
75        873
         ... 
840         1
306         1
2,695       1
2,520       1
291         1
Name: count, Length: 511, dtype: int64
count    30478.000000
mean       163.589737
std        197.785454
min         10.000000
25%         80.000000
50%        125.000000
75%        195.000000
max      10000.000000
Name: Price, dtype: float64


In [84]:
print(airbnb['Beds'].unique())
airbnb['Beds'] = airbnb['Beds'].fillna(1)

[ 1.  3.  2.  4.  5. nan  6. 10.  7.  8. 12.  0. 16.  9. 11.]


In [85]:
print(airbnb['N_Reviews'].unique())

[  0   1  39   4   9  80  95  23  14 120  81  17  32  52   3 171  16  19
   2  28  62   6   5  12  40  47  13  35  34  68  42   7  41  15  21   8
  10  29 112  25 132  51 156  30  98  24  74  20 188 221  57  11 242  67
 118  60 136 119  79 106  43  91 105  96 141 146  63  58  18 150  38  55
  89  46  77  48  36  69  26  73  53 165  92  87 108 109  50 127  88  83
 184 179  22  31  82  27 123 247 190 257 217  85 103 157 102  56 125 115
  49  97  54  65  90  44  33 100  37 107 114 116  61 104  71  75 145  84
  93 110 172 148 131 154  86  64  76 129  45 180 140  78 200 175 189 170
 187  72 137  99 101  59  94 124 223 173 151 142 163 121 256 248 149 128
 133  66 117 176 111  70 183 218 192 206 161 216 191 213 178 122 153 177
 167 113 126 135 181 138 143 195 182 198 211 166 139 134 144 174 159 155
 203 201 164 212 236 235 130]


In [87]:
print(airbnb['Score'].unique())
print(airbnb.shape)

[ nan  96. 100.  94.  90.  98.  93.  91.  97.  95.  99.  85.  86.  80.
  88.  92.  89.  82.  87.  81.  76.  78.  83.  66.  84.  72.  79.  60.
  40.  62.  74.  77.  50.  71.  75.  73.  69.  65.  68.  70.  67.  64.
  20.  57.  58.  43.  63.  55.  47.  53.  49.  30.]
(30478, 4)


In [90]:
airbnb = airbnb.dropna(subset=["Score"])
print(airbnb['Score'].unique())
print(airbnb.shape)

[ 96. 100.  94.  90.  98.  93.  91.  97.  95.  99.  85.  86.  80.  88.
  92.  89.  82.  87.  81.  76.  78.  83.  66.  84.  72.  79.  60.  40.
  62.  74.  77.  50.  71.  75.  73.  69.  65.  68.  70.  67.  64.  20.
  57.  58.  43.  63.  55.  47.  53.  49.  30.]
(22155, 4)


In [91]:
airbnb = airbnb.dropna()

2. Maxmin normalize the data and remove any `nan`'s (`KMeans` from `sklearn` doesn't accept `nan` input).

In [92]:
def maxmin(df):
    df = (df-min(df))/(max(df)-min(df))
    return df

airbnb = airbnb.apply(maxmin)

3. Use `sklearn`'s `KMeans` module to cluster the data by `Beds`, `Number of Reviews`, and `Review Scores Rating` for `k=6`.

In [94]:
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

data = airbnb.loc[:, ['Beds', 'N_Reviews', 'Score']]
print(data)
model = KMeans(n_clusters=6, max_iter=300, n_init = 10, random_state=0) # Create a model for 
model = model.fit(data) # Fit the emodel
data['cluster'] = model.labels_

         Beds  N_Reviews  Score
4      0.1875   0.148438  0.950
5      0.0625   0.011719  1.000
6      0.0625   0.031250  1.000
7      0.0625   0.308594  0.925
8      0.1250   0.367188  0.875
...       ...        ...    ...
30332  0.0625   0.000000  1.000
30347  0.0625   0.000000  0.750
30378  0.0625   0.000000  1.000
30404  0.0625   0.000000  1.000
30409  0.0625   0.000000  0.750

[22155 rows x 3 columns]


AttributeError: 'NoneType' object has no attribute 'split'

4. Use `seaborn`'s `.pairplot()` to make a grid of scatterplots that show how the clustering is carried out in multiple dimensions.

5. Use `.groupby` and `.describe` to compute the average price for each cluster. Which clusters have the highest rental prices?

6. Use a scree plot to pick the number of clusters and repeat steps 4 and 5.

**Q2.** This is a question about $k$ means clustering. We want to investigate how adjusting the "noisiness" of the data impacts the quality of the algorithm and the difficulty of picking $k$.

1. Run the code below, which creates four datasets: `df0_125`, `df0_25`, `df0_5`, `df1_0`, and `df2_0`. Each data set is created by increasing the amount of `noise` (standard deviation) around the cluster centers, from `0.125` to `0.25` to `0.5` to `1.0` to `2.0`.

```
import numpy as np
import pandas as pd

def createData(noise,N=50):
    np.random.seed(100) # Set the seed for replicability
    # Generate (x1,x2,g) triples:
    X1 = np.array([np.random.normal(1,noise,N),np.random.normal(1,noise,N)])
    X2 = np.array([np.random.normal(3,noise,N),np.random.normal(2,noise,N)])
    X3 = np.array([np.random.normal(5,noise,N),np.random.normal(3,noise,N)])
    # Concatenate into one data frame
    gdf1 = pd.DataFrame({'x1':X1[0,:],'x2':X1[1,:],'group':'a'})
    gdf2 = pd.DataFrame({'x1':X2[0,:],'x2':X2[1,:],'group':'b'})
    gdf3 = pd.DataFrame({'x1':X3[0,:],'x2':X3[1,:],'group':'c'})
    df = pd.concat([gdf1,gdf2,gdf3],axis=0)
    return df

df0_125 = createData(0.125)
df0_25 = createData(0.25)
df0_5 = createData(0.5)
df1_0 = createData(1.0)
df2_0 = createData(2.0)
```

2. Make scatterplots of the $(X1,X2)$ points by group for each of the datasets. As the `noise` goes up from 0.125 to 2.0, what happens to the visual distinctness of the clusters?
3. Create a scree plot for each of the datasets. Describe how the level of `noise` affects the scree plot (particularly the presence of a clear "elbow") and your ability to definitively select a $k$.
4. Explain the intuition of the elbow, using this numerical simulation as an example.

**Q3.** We looked at computer vision with $k$NN in a previous question. Can $k$ means clustering correctly group digits, even if we don't know which symbols are which?

1. To load the data, run the following code in a chunk:
```
from keras.datasets import mnist
df = mnist.load_data('minst.db')
train,test = df
X_train, y_train = train
X_test, y_test = test
```
The `y_test` and `y_train` vectors, for each index `i`, tell you want number is written in the corresponding index in `X_train[i]` and `X_test[i]`. The value of `X_train[i]` and `X_test[i]`, however, is a 28$\times$28 array whose entries contain values between 0 and 256. Each element of the matrix is essentially a "pixel" and the matrix encodes a representation of a number. To visualize this, run the following code to see the first ten numbers:
```
import matplotlib.pyplot as plt
import numpy as np
np.set_printoptions(edgeitems=30, linewidth=100000)
for i in range(5):
    print(y_test[i],'\n') # Print the label
    print(X_test[i],'\n') # Print the matrix of values
    plt.contourf(np.rot90(X_test[i].transpose())) # Make a contour plot of the matrix values
    plt.show()
```
OK, those are the data: Labels attached to handwritten digits encoded as a matrix.

2. What is the shape of `X_train` and `X_test`? What is the shape of `X_train[i]` and `X_test[i]` for each index `i`? What is the shape of `y_train` and `y_test`?
3. Use Numpy's `.reshape()` method to covert the training and testing data from a matrix into an vector of features. So, `X_test[index].reshape((1,784))` will convert the $index$-th element of `X_test` into a $28\times 28=784$-length row vector of values, rather than a matrix. Turn `X_train` into an $N \times 784$ matrix $X$ that is suitable for scikit-learn's kNN classifier where $N$ is the number of observations and $784=28*28$ (you could use, for example, a `for` loop).
4. Use $k$ means clustering on the reshaped `X_test` data with `k=10`.  
5. Cross tabulate the cluster assignments with the true labels for the test set values. How good is the correspondence? What proportion of digits are clustered correctly? Which digits are the hardest to distinguish from one another? Can $k$MC recover the latent digits 0 to 9, without even knowing what those digits were?
6. If you use a scree plot to determine the number of clusters $k$, does it pick 10 (the true number of digits), or not? If it fails to pick $k=10$, which digits does it tend to combine into the same classification?